pkg/cloudprovider/providers/azure/azure_util_vmss.go

```diff
@@ -40,124 +42,235 @@ var (
+		vms, ok := ss.cache[scaleSetName]
+		if !ok {
+			glog.V(4).Infof("scale set (%s) not found", scaleSetName)
+			return scaleSetVMInfo{}, cloudprovider.InstanceNotFound
+		}
 
-	// scaleSetName is required to query VM info.
-	if scaleSetName == "" {
-		scaleSetName = ss.PrimaryScaleSetName
+		for _, vm := range vms {
+			if vm.InstanceID == instanceID {
+				glog.V(4).Infof("getCachedVirtualMachineByInstanceID gets vm (%s) by instanceID (%s) within scale set (%s)", vm.NodeName, instanceID, scaleSetName)
+				return vm, nil
+			}
+		}
+
+		glog.V(4).Infof("instanceID (%s) not found in scale set (%s)", instanceID, scaleSetName)
+		return scaleSetVMInfo{}, cloudprovider.InstanceNotFound
 	}
 
-	ss.operationPollRateLimiter.Accept()
-	glog.V(10).Infof("VirtualMachineScaleSetVMsClient.Get(%s): start", instanceID)
-	vm, err = ss.VirtualMachineScaleSetVMsClient.Get(ss.ResourceGroup, scaleSetName, instanceID)
-	glog.V(10).Infof("VirtualMachineScaleSetVMsClient.Get(%s): end", instanceID)
+	vm, err := getVMByID(scaleSetName, instanceID)
+	if err == nil {
+		return vm, nil
+	}
 
-	exists, realErr = checkResourceExistsFromError(err)
-	if realErr != nil {
-		return vm, false, realErr
+	// Update cache and try again.
+	if err = ss.updateCache(); err != nil {
+		return scaleSetVMInfo{}, err
+	}
+	vm, err = getVMByID(scaleSetName, instanceID)
+	if err == nil {
+		return vm, nil
 	}
 
-	if !exists {
-		return vm, false, nil
+	return scaleSetVMInfo{}, cloudprovider.InstanceNotFound
+}
+
+// GetInstanceIDByNodeName gets the cloud provider ID by node name.
+// It must return ("", cloudprovider.InstanceNotFound) if the instance does
+// not exist or is no longer running.
+func (ss *scaleSet) GetInstanceIDByNodeName(name string) (string, error) {
+	vm, err := ss.getCachedVirtualMachine(name)
+	if err != nil {
+		if err == cloudprovider.InstanceNotFound {
+			glog.V(4).Infof("GetInstanceIDByNodeName: node %q is not found in scale sets, assuming it is managed by availability set", name)
+
+			// Retry with standard type because master nodes may not belong to any vmss.
+			// TODO: find a better way to identify the type of VM.
+			return ss.availabilitySet.GetInstanceIDByNodeName(name)
+		}
+
+		return "", err
 	}
 
-	return vm, exists, err
+	return vm.ID, nil
 }
 
 // GetNodeNameByProviderID gets the node name by provider ID.
 func (ss *scaleSet) GetNodeNameByProviderID(providerID string) (types.NodeName, error) {
 	// NodeName is not part of providerID for vmss instances.
-	parts := strings.Split(providerID, "/")
-	instanceID := parts[len(parts)-1]
-	machine, exist, err := ss.getScaleSetVMByID(instanceID, ss.PrimaryScaleSetName)
-	if !exist {
-		return "", cloudprovider.InstanceNotFound
+	scaleSetName, err := extractScaleSetNameByVMID(providerID)
+	if err != nil {
+		glog.V(4).Infof("Can not extract scale set name from providerID (%s), assuming it is mananaged by availability set", providerID)
```

Comment:
Infof ? Pls print err, and it does not return err?

Reply:
Fixed.